src/Language/PureScript/Docs/Convert.hs

```diff
@@ -71,6 +86,65 @@ insertValueTypes env m =
       Nothing ->
         err ("name not found: " ++ show key)
 
+  toKindSignatureFor :: P.DataDeclType -> P.KindSignatureFor
+  toKindSignatureFor = \case
+    P.Data -> P.DataSig
+    P.Newtype -> P.NewtypeSig
+
+  insertInferredKind :: Declaration -> Text -> P.KindSignatureFor -> Declaration
+  insertInferredKind d name keyword =
+    let
+      key = P.Qualified (Just (modName m)) (P.ProperName name)
+    in case Map.lookup key (P.types env) of
+      Just (inferredKind, _) ->
+        if isFunctionTypeStarTerminal inferredKind'
+          then  d
+          else  d { declKind = Just $ KindInfo
+                    { kiKeyword = keyword
+                    , kiKind = dropTypeSortAnnotation inferredKind'
+                    }
+                  }
+        where
+          inferredKind' = inferredKind $> ()
+          kindPrimType = P.TypeConstructor () Prim.Type
+          kindPrimFunction = P.TypeConstructor () Prim.Function
+          kindPrimConstraint = P.TypeConstructor () Prim.Constraint
+
+          -- `Type ->`
+          kindFunctionType = P.TypeApp () kindPrimFunction kindPrimType
+
+          -- |
+          -- Returns True if the inferred kind signature follows this
+          -- pattern: `(Type ->)* [Terminal]` where `Terminal` is
+          -- `Type` or `Constraint`
+          --
+          -- Concretely, these will return True
+          -- - Type
+          -- - Constraint
+          -- - Type -> Type
+          -- - Type -> Constraint
+          -- - Type -> Type -> ... -> Type
+          -- - Type -> Type -> ... -> Constraint
+          isFunctionTypeStarTerminal
+            :: P.Type () -> Bool
+          isFunctionTypeStarTerminal = \case
+            P.TypeApp _ t1 t2 | t1 == kindFunctionType ->
+              isFunctionTypeStarTerminal t2
+            x
+              | x == kindPrimType || x == kindPrimConstraint -> True
```

Comment:
The `Constraint` thing is a good point; I hadn't considered that classes have a different ‘most intuitive’ kind.

This `||` is correct because there's no way to make something `Constraint`-kinded that isn't a class right now, right? That is, something like
```purs
type IntState = MonadState Int
```
currently isn't supported, and there isn't anything else like that which could create a `Constraint` other than a class?

I'd feel a tiny bit more comfortable if this `||` were instead conditional on the declaration type—a `x == if isTypeClass d then kindPrimConstraint else kindPrimType` sort of deal—because some day type synonyms for constraints might be a thing? But perhaps that's unnecessary future-proofing, if we know for sure that there aren't any `Constraint` loopholes in the present.

Reply:
That's a good suggestion - I think full constraint kinds (which includes being able to write type synonyms like your `IntState` one) is something that's likely to happen in the future, so I agree about making this conditional on the declaration type.

Reply:
https://github.com/purescript/purescript/issues/1764